File: apidiff/declarations.mjs
/**
 * declarations.mjs — extract C# declarations from a DocFX type page (Node port of declarations.py).
 *
 * Every declaration is anchored to the `data-uid` on its heading, which DocFX writes byte-identical
 * to the xrefmap `uid`, so no href/anchor demangling is needed. The type's own declaration hangs off
 * its `<h1 data-uid>`. This layout is unchanged from the oldest archived pages (2022) to the newest.
 *
 * Anchoring is a correctness requirement, not a style preference: Archetype-N injects *code samples*
 * into these same pages, so a flat scan for `lang-csharp` blocks — the approach in
 * scaffolds/tools/archetype-n/facts.mjs — reports `using IronZip;` as a member.
 */

import { stripGuidMarkers } from "./xrefmap.mjs";

/** Archetype-N injects prose and runnable samples between these sentinels; removed before parsing. */
const ARCHETYPE_BLOCK = /<!--\s*archetype-N:start[\s\S]*?<!--\s*archetype-N:end\s*-->/gi;

/** A heading carrying a uid. The `<a data-uid="...*">` overload anchors are not headings. */
const HEADING_WITH_UID = /<h[1-6][^>]*\sdata-uid="([^"]+)"[^>]*>/gi;

/** One rendered declaration. DocFX emits `lang-csharp hljs`; the suffix is allowed to vary. */
const CSHARP_BLOCK = /<code class="lang-csharp[^"]*">([\s\S]*?)<\/code>/gi;

const TAG = /<[^>]+>/g;

/**
 * The "Implements" block a type page carries above its Syntax heading, e.g.
 *
 *     <h5>Implements</h5>
 *         <div><span class="xref">System.Collections.Generic.IEnumerable</span>&lt;…Cell…&gt;</div>
 *         <div><span class="xref">System.Collections.IEnumerable</span></div>
 *
 * This is the authoritative record of the interfaces a type implements, and it is *stable across
 * DocFX versions* — unlike the declaration line, which stopped inlining interfaces between the
 * 2026.6 and 2026.7 builds while this section stayed byte-identical.
 */
const IMPLEMENTS_SECTION = /<h5>\s*Implements\s*<\/h5>([\s\S]*?)(?=<h[1-6])/i;
const IMPLEMENTS_ENTRY = /<div>([\s\S]*?)<\/div>/g;

/** A dotted, fully-qualified name; reduced to its last segment to match the declaration's form. */
const QUALIFIED_NAME = /[A-Za-z_][A-Za-z0-9_]*(?:\.[A-Za-z_][A-Za-z0-9_]*)+/g;

/**
 * Strip namespaces from every qualified name in a type expression, so
 * `IronSoftware.Abstractions.IParent<IronSoftware.Abstractions.Word.IWordDocumentObjectCollection>`
 * becomes `IParent<IWordDocumentObjectCollection>`.
 */
export function simplifyTypeName(text) {
  return text.replace(QUALIFIED_NAME, (match) => match.slice(match.lastIndexOf(".") + 1));
}

/** Minimal HTML entity decoding — the set DocFX actually emits inside declarations. */
const ENTITIES = { amp: "&", lt: "<", gt: ">", quot: '"', apos: "'", "#39": "'", nbsp: " " };

function unescapeHtml(text) {
  return text.replace(/&(#x?[0-9a-fA-F]+|[a-zA-Z]+);/g, (match, name) => {
    if (Object.prototype.hasOwnProperty.call(ENTITIES, name)) return ENTITIES[name];
    if (name.startsWith("#x") || name.startsWith("#X")) return String.fromCodePoint(parseInt(name.slice(2), 16));
    if (name.startsWith("#")) return String.fromCodePoint(parseInt(name.slice(1), 10));
    return match;
  });
}

/**
 * Turn a raw declaration code block into a single comparable line.
 *
 * GUID markers are stripped here too, so a declaration mentioning a marked type stays comparable
 * across builds — and matches the uid, which xrefmap.mjs strips the same way.
 */
export function normalizeDeclaration(raw) {
  return stripGuidMarkers(unescapeHtml(raw.replace(TAG, "")).replace(/\s+/g, " ").trim());
}

/**
 * Map every `data-uid` on a DocFX type page to its C# declaration.
 *
 * Includes the type's own uid (from its `<h1>`). Uids with no declaration between their heading and
 * the next are omitted rather than recorded as empty.
 *
 * @param {string} pageHtml Full text of an `api/<Type>.html` page.
 * @returns {Object<string,string>} `{ [uid]: declaration }`
 */
export function parseTypePage(pageHtml) {
  const cleaned = pageHtml.replace(ARCHETYPE_BLOCK, "");

  // Overload-group uids end in `*` and describe a set of overloads, not a signature.
  const anchors = [];
  for (const match of cleaned.matchAll(HEADING_WITH_UID)) {
    if (!match[1].endsWith("*")) anchors.push([match.index, stripGuidMarkers(match[1])]);
  }
  if (anchors.length === 0) return {};

  const blocks = [...cleaned.matchAll(CSHARP_BLOCK)].map((match) => [match.index, match[1]]);
  if (blocks.length === 0) return {};

  const declarations = Object.create(null);
  for (let index = 0; index < anchors.length; index++) {
    const [position, uid] = anchors[index];
    // A declaration belongs to the nearest preceding heading, so bound the search at the next one.
    const nextPosition = index + 1 < anchors.length ? anchors[index + 1][0] : cleaned.length;
    for (const [blockPosition, blockText] of blocks) {
      if (blockPosition < position) continue;
      if (blockPosition >= nextPosition) break;
      // The first block in a section is the Declaration; later ones are Examples.
      const declaration = normalizeDeclaration(blockText);
      if (declaration) declarations[uid] = declaration;
      break;
    }
  }
  return declarations;
}

/** Return the interfaces listed in a type page's Implements section, simple-named and sorted. */
export function parseImplements(pageHtml) {
  const section = IMPLEMENTS_SECTION.exec(pageHtml);
  if (!section) return [];
  const interfaces = new Set();
  for (const entry of section[1].matchAll(IMPLEMENTS_ENTRY)) {
    const name = normalizeDeclaration(entry[1]);
    if (name) interfaces.add(simplifyTypeName(name));
  }
  return [...interfaces].sort((a, b) => (a < b ? -1 : a > b ? 1 : 0));
}

File: apidiff/declarations.py
"""Extract C# declarations from a generated DocFX type page — the signature layer of a diff.

Every declaration is anchored to the ``data-uid`` on its heading, which DocFX writes byte-identical
to the xrefmap ``uid``:

    <h4 id="IronZip_IronZipArchive_Contains_System_String_"
        data-uid="IronZip.IronZipArchive.Contains(System.String)">Contains(String)</h4>
    <h5 class="decalaration">Declaration</h5>
    <div class="codewrapper">
      <pre><code class="lang-csharp hljs">public bool Contains(string EntryName)</code></pre>
    </div>

so no href/anchor demangling is needed. The type's own declaration hangs off its ``<h1 data-uid>``.
This layout is unchanged from the oldest archived pages (2022) to the newest.

Anchoring is a correctness requirement, not a style preference: Archetype-N injects *code samples*
into these same pages, so a flat scan for ``lang-csharp`` blocks — the approach in
``scaffolds/tools/archetype-n/facts.py`` — reports ``using IronZip;`` as a member.
"""

import html
import re

from .xrefmap import strip_guid_markers

# Archetype-N injects prose and runnable samples between these sentinels. Removed before parsing so
# an injected sample can never be mistaken for a declaration.
ARCHETYPE_BLOCK = re.compile(r"<!--\s*archetype-N:start.*?<!--\s*archetype-N:end\s*-->", re.DOTALL | re.I)

# A heading carrying a uid. The `<a data-uid="...*"></a>` overload-group anchors are not headings and
# so are not matched here; the trailing-`*` guard below is a second line of defence.
HEADING_WITH_UID = re.compile(r"<h[1-6][^>]*\sdata-uid=\"([^\"]+)\"[^>]*>", re.I)

# One rendered declaration. DocFX emits `lang-csharp hljs`; the suffix is allowed to vary.
CSHARP_BLOCK = re.compile(r"<code class=\"lang-csharp[^\"]*\">(.*?)</code>", re.DOTALL | re.I)

# Inline markup can appear inside a declaration block (cross-reference links on type names).
TAG = re.compile(r"<[^>]+>")


def normalize_declaration(raw: str) -> str:
    """Turn a raw declaration code block into a single comparable line.

    GUID markers are stripped here too, so a declaration mentioning a marked type stays comparable
    across builds — and matches the uid, which xrefmap.py strips the same way.
    """
    return strip_guid_markers(re.sub(r"\s+", " ", html.unescape(TAG.sub("", raw))).strip())


def parse_type_page(page_html: str) -> dict:
    """Map every ``data-uid`` on a DocFX type page to its C# declaration.

    Args:
        page_html (str): Full text of an ``api/<Type>.html`` page.

    Returns:
        dict: ``{uid: declaration}``. Includes the type's own uid (from its ``<h1>``). Uids with no
        declaration between their heading and the next are omitted rather than recorded as empty.
    """
    cleaned = ARCHETYPE_BLOCK.sub("", page_html)

    # Overload-group uids end in `*` and describe a set of overloads, not a signature.
    anchors = [
        (match.start(), strip_guid_markers(match.group(1)))
        for match in HEADING_WITH_UID.finditer(cleaned)
        if not match.group(1).endswith("*")
    ]
    if not anchors:
        return {}

    blocks = [(match.start(), match.group(1)) for match in CSHARP_BLOCK.finditer(cleaned)]
    if not blocks:
        return {}

    declarations = {}
    for index, (position, uid) in enumerate(anchors):
        # A declaration belongs to the nearest preceding heading, so bound the search at the next one.
        next_position = anchors[index + 1][0] if index + 1 < len(anchors) else len(cleaned)
        for block_position, block_text in blocks:
            if block_position < position:
                continue
            if block_position >= next_position:
                break
            # The first block in a section is the Declaration; later ones are Examples.
            declaration = normalize_declaration(block_text)
            if declaration:
                declarations[uid] = declaration
            break

    return declarations
